codemodder/codemods/transformations/clean_imports.py

```diff
@@ -57,148 +57,272 @@ def __init__(self, context: CodemodContext, src_path):
+            replacement_nodes = replacement_nodes | {node: RemovalSentinel.REMOVE}
+        result_tree = result_tree.visit(ReplaceNodes(replacement_nodes))
+        return result_tree
+
+    def _split_leading_lines(self, stmt: cst.SimpleStatementLine) -> int:
+        """
+        For a given SimpleStatementLine, split the leading_lines nodes head and tail, where tail will contain the longest comment block right before the line.
+        """
+        i = len(stmt.leading_lines) - 1
+        while i >= 0:
+            empty_line = stmt.leading_lines[i]
+            if not matchers.matches(
+                empty_line, matchers.EmptyLine(comment=matchers.Comment())
+            ):
+                break
+            i = i - 1
+        return i + 1
+
+    def _trim_leading_lines(self, stmt: cst.SimpleStatementLine) -> List[cst.EmptyLine]:
+        """
+        For a given SimpleStatementLine, gather the list of comments right above it.
+        """
+        i = self._split_leading_lines(stmt)
+        return list(stmt.leading_lines[i:])
+
+    def _handle_regular_imports(self, import_stmts):
+        # bin them into (name,alias) for deduplication and comment merging
+        # the value of the dict contains the comment blocks from all the same imports
+        alias_dict = defaultdict(list)
+        for stmt in import_stmts:
+            imp = stmt.body[0]
+            comments = self._trim_leading_lines(stmt)
+            # only the first import in a list will inherit the comments
+            first_ia = imp.names[0]
+            alias_dict.setdefault(
+                (first_ia.evaluated_name, first_ia.evaluated_alias), []
+            )
+            if comments:
+                alias_dict[(first_ia.evaluated_name, first_ia.evaluated_alias)].append(
+                    comments
+                )
+            for ia in imp.names[1:]:
+                alias_dict.setdefault((ia.evaluated_name, ia.evaluated_alias), [])
+
+        # create the statements for each pair in alias_dict
+        new_import_stmts = []
+        for name, asname in alias_dict.keys():
+            ia = cst.ImportAlias(
+                name=_node_from_name(name),
+                asname=(asname and cst.AsName(cst.Name(asname))),
+            )
+            # invert the comment blocks and flatten
+            comments = list(
+                itertools.chain.from_iterable(reversed(alias_dict[(name, asname)]))
+            )
+            new_import_stmts.append(
+                cst.SimpleStatementLine(
+                    body=[cst.Import(names=[ia])], leading_lines=comments
+                )
+            )
+        return new_import_stmts
+
+    def _handle_from_imports(self, from_import_stmts):
+        imports_dict = defaultdict(lambda: (set(), []))
+        star_imports_dict = defaultdict()
+        # binsort the aliases into module name, merge comments
+        for stmt in from_import_stmts:
```

Comment:
Just a small suggestion but this function might read a little cleaner if each of these loops was broken into a separate method. Each comment above these loops suggests a new method name.